[RISCV][llvm-readobj,llvm-objdump] Use getRISCVVendorRelocationTypeName for RISCV vendor relocations

kito-cheng opened the pull request:
Use getRISCVVendorRelocationTypeName to resolve RISCV vendor-specific relocation names (R_RISCV_CUSTOM192-255) when preceded by R_RISCV_VENDOR.

This improves the output of llvm-readobj and llvm-objdump to show vendor-specific names like R_RISCV_QC_ABS20_U, R_RISCV_QC_E_BRANCH (QUALCOMM) and R_RISCV_NDS_BRANCH_10 (ANDES) instead of generic R_RISCV_CUSTOM* names.

Per RISC-V psABI, R_RISCV_VENDOR must be placed immediately before its associated vendor-specific relocation, so the vendor symbol is consumed after one use. Unknown vendors fall back to R_RISCV_CUSTOM*.

kito-cheng committed:
[RISCV][llvm-readobj,llvm-objdump] Use getRISCVVendorRelocationTypeName for RISCV vendor relocations

Use getRISCVVendorRelocationTypeName to resolve RISCV vendor-specific
relocation names (R_RISCV_CUSTOM192-255) when preceded by R_RISCV_VENDOR.

This improves the output of llvm-readobj and llvm-objdump to show
vendor-specific names like R_RISCV_QC_ABS20_U, R_RISCV_QC_E_BRANCH
(QUALCOMM) and R_RISCV_NDS_BRANCH_10 (ANDES) instead of generic
R_RISCV_CUSTOM* names.

Per RISC-V psABI, R_RISCV_VENDOR must be placed immediately before its
associated vendor-specific relocation, so the vendor symbol is consumed
after one use. Unknown vendors fall back to R_RISCV_CUSTOM*.

kito-cheng committed:
!fixup format code

jrtc27 commented on llvm/test/tools/llvm-objdump/ELF/RISCV/riscv-vendor-reloc.s:
```diff
@@ -0,0 +1,64 @@
+  .reloc ., R_RISCV_VENDOR, QUALCOMM
+  .reloc ., R_RISCV_CUSTOM193, quux
+  nop
+
```
```
## Test 6: An unpaired R_RISCV_VENDOR followed by an R_RISCV_CUSTOM* at a different offset
  .reloc . - 1, R_RISCV_VENDOR, QUALCOMM
  .reloc ., R_RISCV_CUSTOM193, barney

## Test 7: A non-R_RISCV_CUSTOM* in between a vendor relocation pair
  .reloc ., R_RISCV_VENDOR, QUALCOMM
  .reloc ., R_RISCV_32, snork
  .reloc ., R_RISCV_CUSTOM193, zot
```

I believe those are currently handled incorrectly, both will be treated as valid vendor relocation pairs but neither are.

kito-cheng committed:
!fixup validate R_RISCV_VENDOR pairing requires same offset and immediate adjacency

Per RISC-V psABI, R_RISCV_VENDOR must be placed immediately before the
vendor-specific relocation at the same offset. This fixes two edge cases:
- Different offsets between R_RISCV_VENDOR and R_RISCV_CUSTOM*
- Non-CUSTOM relocation in between breaks the pairing

kito-cheng commented:
ping :)

lenary reviewed (approved):
LGTM. Thanks for covering this!
